### PR TITLE
Refactor search SQL operations into helper class, catalog side.

### PR DESCRIPTION
**Pull request by neekfenwick**
Took the bulk of search_result/header.php which built the SQL and put it in a helper class Zencart\Search\Search.
Inputs to the search refactored into SearchOptions so they can be called programatically without needing a GET request to trigger it. Parsing of $_GET retained in the search/header.php and search_result/header.php.

New message_stack: 'search_result' for displaying a message at the top of
  the search results page (for example 'try also doing xyz').
New notifier: NOTIFY_SEARCH_NO_RESULTS_MESSAGE, given $results, $message
  For reacting to the search results e.g. add to message stack (to appear
  on the 'search' form).

### CODE REVIEW

File: includes/auto_loaders/config.core.php
Comment (proseLA) [inline includes/auto_loaders/config.core.php:75]: i see no reason why you should not use the short array syntax

File: includes/auto_loaders/config.core.php
Comment (proseLA) [inline includes/auto_loaders/config.core.php:175]: same comment as above with regards to the short array syntax

File: includes/modules/pages/search_result/header_php.php
Comment (proseLA) [inline includes/modules/pages/search_result/header_php.php:8]: it looks like you are making a PR against the v1.5.8-alpha tag.  i believe all PRs should be going against the master branch.

IMO, changes in the stamp should only be done via the repo maintainers, and then when they are doing some sort of new tag or release.  otherwise it can get very confusing when doing upgrades.  if a PR changes the stamp, i would surmise something is a bit adrift with your process.

**Review by proseLA — CHANGES_REQUESTED** [review]
big fan of moving this into a class file.

File: includes/classes/class.search.php
Comment (proseLA) [inline includes/classes/class.search.php:60]: this class file has some notifiers in it.  have you tested them? 

i believe this class file needs to extend base in order for those notifiers to work.

File: includes/auto_loaders/config.core.php
Comment (neekfenwick) [reply includes/auto_loaders/config.core.php:75]: Fair enough, I was trying to fit in with old code but I see zcDate's changes use the new syntax, and tighter indentation. I've updated my syntax.

File: includes/classes/class.search.php
Comment (neekfenwick) [reply includes/classes/class.search.php:60]: Good catch, I had rather stupidly imported the global `$zco_notifier` instead of deriving from `base` and using `$this->notify` .. this is now fixed.  I have tested with an observer and it is fired correctly.

File: includes/modules/pages/search_result/header_php.php
Comment (neekfenwick) [reply includes/modules/pages/search_result/header_php.php:8]: I thought this work was development on v158, sorry.  I have rebased my change onto upstream/master and force pushed (I believe a force push is required in this case, a simple push was rejected by git).

**Comment by neekfenwick** [comment]
Note to maintainer: This branch contained multiple commits; I ran into trouble earlier with my FontAwesome PR when you merged the branch plainly leaving multiple commits on `master`.  I have rebased this branch to squash my commits into a single one, I hope this helps.

File: includes/modules/pages/search_result/header_php.php
Comment (proseLA) [reply includes/modules/pages/search_result/header_php.php:8]: no apologies necessary.

unfortunately, this still looks wrong.  i would refer to #5911 with regards to how PRs should work.

with regards to the `$Id` and other elements in the stamp, how they work is not important (at least to me).  i have opened issues and there have been discussions on them in the past, and i am comfortable with whatever process the maintainers use.  the point that i am making is that if your PR changes the stamp (and yours does), then there is a problem with your process of creating that PR.  again, i would refer to the above mentioned post.

however, in doing a cursory look, it does seem that your new class file, when using existing notifiers, those notifiers have the correct signature, ie they are the same as in the master branch; and not what is in the v1.5.8a tag.

as an example:

```
charmes: # git checkout v1.5.8a-alpha 
Note: switching to 'v1.5.8a-alpha'.

You are in 'detached HEAD' state
....
charmes: # git diff master includes/modules/pages/search_result/header_php.php
...
// Notifier Point
-$zco_notifier->notify('NOTIFY_SEARCH_COLUMNLIST_STRING', $select_column_list, $select_column_list);
+$zco_notifier->notify('NOTIFY_SEARCH_COLUMNLIST_STRING');

...
 
 // Notifier Point
-$zco_notifier->notify('NOTIFY_SEARCH_SELECT_STRING', $select_str, $select_str);
+$zco_notifier->notify('NOTIFY_SEARCH_SELECT_STRING');

```
in the above section of code, the added parameter elements to the notifiers are in the master branch, while the notifiers without parameters on in the v1.5.8a-alpha tag.  and when i look at your new class file for those notifiers, the parameters are there:

https://github.com/zencart/zencart/blob/86e2bd9690e46aaa9e61e26b88992ec03f592024/includes/classes/class.search.php#L198

https://github.com/zencart/zencart/blob/86e2bd9690e46aaa9e61e26b88992ec03f592024/includes/classes/class.search.php#L210

i'm not sure what more to say on it...

hopefully it makes sense to you on it.  and again, i am a fan of what you are doing here.

Comment (neekfenwick) [reply includes/modules/pages/search_result/header_php.php:8]: I'm a little confused as to how those lines are wrong.  Github 'show files' on my PR shows, for example, the COLUMNLIST_STRING one looks like (from https://github.com/zencart/zencart/pull/6036/files) :

`        $this->notify('NOTIFY_SEARCH_COLUMNLIST_STRING', $select_column_list, $select_column_list);`

This is correct, no?  If you go back to the v1.5.8.a-alpha tag then the two parameters are missing, yes, but that's old code and was improved apparently by yourself at https://github.com/zencart/zencart/commit/fd0820894fbc1173e705bb515e625a9f29653524

Visible in latest code on `master`: https://github.com/zencart/zencart/blob/9ae0b145fff3d89bbeafdae83b1f9c00dbdf6788/includes/modules/pages/search_result/header_php.php#L211

Can you spell out what's wrong in a little more detail please?

Comment (proseLA) [reply includes/modules/pages/search_result/header_php.php:8]: yes, the code looks correct.  (i have not tested it.)

you based your changes on the wrong branch; which is why the header comments have changed.  somehow you seem to have done the code correctly.  the commit you referenced, fd08208, was done 11/23/2022.  and yet the header that you are now changing is back to a release on 7/13/2022.

if you are confident that the code is correct, and no one else has a problem, i would not worry about it.  

if you still do no understand what i am saying, perhaps someone else can help clarify it.

Comment (neekfenwick) [reply includes/modules/pages/search_result/header_php.php:8]: Well, "check twice, merge once" as someone somewhere probably says... Upon diffing against the original `search_results/php_header.php` I've found a couple of places `$_GET` was still used in the refactored code.  I've fixed these up to use the `$searchOptions` object as they should, but not committed or pushed yet, because of the next point.

I also note, the new class.search.php file has no Copyright comment at the top or place for the `$Id blah$` string. @drbyte could you please say what I should put in a brand new file?  Is it enough to have something like:

```
/**
 * @copyright Copyright 2003-2023 Zen Cart Development Team
 * @license http://www.zen-cart.com/license/2_0.txt GNU Public License V2.0
 * @version $Id:  $
 */
```

and something will fill in the `$Id$` bit automatically?  I've always seen people's username, a date and some kind of tag id in there but never known where they come from or how they get updated; I presume it's not manually typed in by someone.

Comment (neekfenwick) [reply includes/modules/pages/search_result/header_php.php:8]: Due to lack of response, I've copied the copyright header from `search_result/header_php.php` (where I refactored most of the code from to make `class.search.php`) and force pushed my branch, making sure it's rebased onto upstream/master.  I presume the `Id` tags will sort themselves out in the merge.

**Comment by neekfenwick** [comment]
I think I want to refactor the code so the `Search` class instance owns its `$searchOptions` as a member, rather than being passed it as a parameter, i.e. from:
```
class SearchOptions {...};
class Search {
    public function buildSearchSQL($searchOptions) {...}
}

```
to

```
class SearchOptions {...};
class Search {
    protected SearchOptions $searchOptions;
    public function __construc(SearchOptions $searchOptions) { $this->$searchOptions = $searchOptions; }
    public function buildSearchSQL() {...}
}
```

In this way, the Observers reacting to notifier calls can access `$callingClass->searchOptions` without needing to be passed it in the notifier arguments.

Background: I've noticed that the old search system used to have the 'search in description' checkbox which was used within the code to add the match against the `pd.products_description` value.   This `$_GET['search_in_description']` value was not expressed as a local variable or passed into any of the notifiers.

Being a fairly straight port of the code, the `Search` class building the SQL also does not pass this value into the notifiers.  Observers need to know all the search options, such as `search_in_description`, to make informed decisions.  In my case, an addon I'm writing wants to add a match against `products_description` but doesn't know if that search option is enabled or not.

I'll update my PR (again! 😀) @proseLA you've gone a bit quiet, are you not happy with the situation with the `Id` tag or is there some other reason this wasn't merged yet?  It's a good thing, as it turns out, as I'm still tweaking it like this.

**Comment by neekfenwick** [comment]
I think I want to refactor the code so the Search class instance owns its $searchOptions as a member, rather than being passed it as a parameter, i.e. from:

```
class SearchOptions {...};
class Search {
    public function buildSearchSQL($searchOptions) {...}
}
```

to

```
class SearchOptions {...};
class Search {
    protected SearchOptions $searchOptions;
    public function setSearchOptions(SearchOptions $searchOptions)
    {
        $this->searchOptions = $searchOptions;
    }
    public function buildSearchSQL() {...}
}
```

In this way, the Observers reacting to notifier calls can access $callingClass->searchOptions without needing to be passed it in the notifier arguments.  Note that we cannot pass `$searchOptions` into the constructor because `Search` is instantiated in the auto_load sequence.  We could pass it into the `buildSearchSQL()` function and immediate set `$this->searchOptions = $searchOptions` but that feels like an antipattern, better to set it in a setter and then have any functions in `Search` that would need it have it available.

I'm keeping the members of `SearchOptions` as `public` so that notifiers will be able to modify them if need be, which could be useful especially from some early stage notifier hooks that want to modify the search process.  Making them `protected` and adding getters/setters seems like overkill.

Background: I've noticed that the old search system used to have the 'search in description' checkbox which was used within the code to add the match against the pd.products_description value. This $_GET['search_in_description'] value was not expressed as a local variable or passed into any of the notifiers.

Being a fairly straight port of the code, the Search class building the SQL also does not pass this value into the notifiers. Observers need to know all the search options, such as search_in_description, to make informed decisions. In my case, an addon I'm writing wants to add a match against products_description but doesn't know if that search option is enabled or not.

I'll update my PR (again! 😀) @proseLA you've gone a bit quiet, are you not happy with the situation with the Id tag or is there some other reason this wasn't merged yet? It's a good thing, as it turns out, as I'm still tweaking it like this.

**Comment by proseLA** [comment]
> I'll update my PR (again! 😀) @proseLA you've gone a bit quiet, are you not happy with the situation with the Id tag or is there some other reason this wasn't merged yet? It's a good thing, as it turns out, as I'm still tweaking it like this.

please do not mistake my lack of response for something that it is not.  i am following along, and when i have some time and energy, i will (hopefully) download and test the PR a bit more thoroughly.

again, i am a fan of what you are doing, as well as the energy that you bring to this project.  you are a most welcome addition.

with regards to the Id tag thing, please, i am NOT upset with that situation at all.  i was pointing that at, more for your own process, than anything else.  that stamp will get changed once the code is merged and a new release is done.  

as someone else has said, stamps are temporary, blame is forever.

finally, please remember, i am just another coder like you.  i try and support this project, but i am not a maintainer of this project.  i have no merge authority.  my opinion is just another voice in the wilderness...  so to speak...

best.

File: includes/languages/lang.english.php
Comment (drbyte) [inline includes/languages/lang.english.php:190]: Should 'sort' be 'search' here? Both in the constant name and its defined value?

File: includes/modules/pages/search_result/header_php.php
Comment (drbyte) [reply includes/modules/pages/search_result/header_php.php:8]: Yes, the basic copyright header which you copied over is fine. The `@version $Id: $` string gets updated when we run the script to stamp all files shortly before tagging a release. Until that stamping it's just the same as the previous stamping.

**Comment by drbyte** [comment]
I like the approach you're taking here.
And I think the embedding of the options makes sense.

File: includes/languages/lang.english.php
Comment (neekfenwick) [reply includes/languages/lang.english.php:190]: Whoops, you're right. Will fix.

**Comment by neekfenwick** [comment]
> please do not mistake my lack of response for something that it is not. i am following along, and when i have some time and energy, i will (hopefully) download and test the PR a bit more thoroughly.

Cheers :) I'm doing all this alongside my day job and it can be easy to go down a psyche rabbit hole sometimes. I'm working on a Soundex Search addon that is a good testing bed for these `Search` changes, but it can be hard to get the time to work through it.  I have a bunch more critical path stuff to deal with this week and will go over the Search changes when possible.

> as someone else has said, stamps are temporary, blame is forever.

Heh :)

**Comment by brittainmark** [comment]
Can anyone forsee a reason why the options would be needed outside the class? If so should a  public function getSearchOptions() be added?

**Comment by neekfenwick** [comment]
> Can anyone forsee a reason why the options would be needed outside the class? If so should a public function getSearchOptions() be added?

Yes, from Observers who want to access the search options.  I can certainly add a public getter for it.

Looks like my mistake - `Search::searchOptions` is currently `protected`, yet as I said in earlier comment, I access `$callingClass->searchOptions->search_in_description` from my Observer which I now notice has been failing silently with a Warning logged.

**Comment by drbyte** [comment]
Looks great!  Thanks for all your work on this!

**Pull request merged**
